### PR TITLE
Add documentation for Fluent migrations

### PR DESCRIPTION
**Pull request by flodolo**

### CODE REVIEW

**Review by Pike — CHANGES_REQUESTED** [review]
Thanks for documenting this. I've got a rather lengthy list of notes, though.

I'd split the explanations from the process. For example, *don't push migrations from quarantine to gecko-strings outside of this process* should be the first thing on the docs ;-)

And then the actual process can be shorter. Hopefully that makes it easier to follow.

More comments inline.

File: src/products/firefox_desktop/fluent_migrations.md
Comment (Pike) [inline src/products/firefox_desktop/fluent_migrations.md]: Actually, `active_scm_l10n`. AFAICT, https://people.mozilla.org/e#nav-access-groups is the least horrible URL to share on how to find out if you have that or not.

File: src/products/firefox_desktop/fluent_migrations.md
Comment (Pike) [inline src/products/firefox_desktop/fluent_migrations.md]: Just `fluent.migrate` is enough, it comes with a package dependency on `compare-locales`. Also makes sure you're installing the right one.

File: src/products/firefox_desktop/fluent_migrations.md
Comment (Pike) [inline src/products/firefox_desktop/fluent_migrations.md]: Filed https://github.com/flodolo/fluent-migrations/issues/3 to make the folder structure a bit more clear. I think that belongs more into the realm of that repo?

File: src/products/firefox_desktop/fluent_migrations.md
Comment (Pike) [inline src/products/firefox_desktop/fluent_migrations.md]: Are there any tips what to do when errors show up?

File: src/products/firefox_desktop/fluent_migrations.md
Comment (flodolo) [reply src/products/firefox_desktop/fluent_migrations.md]: Interesting, is that different from the permissions on the repo? That's `scm_l10n`, or at least that's the one we ask for when requesting new repos.

Didn't even notice those permissions are visible in PMO.

File: src/products/firefox_desktop/fluent_migrations.md
Comment (flodolo) [reply src/products/firefox_desktop/fluent_migrations.md]: Yes, I think so too.

File: src/products/firefox_desktop/fluent_migrations.md
Comment (flodolo) [reply src/products/firefox_desktop/fluent_migrations.md]: I haven't seen Python errors in a long time, because the code is more resilient. It's more frequent to see hg errors. Adding a note anyway.

**Comment by flodolo** [comment]
Thanks for the review, hopefully addressed all points in the next commit.

File: src/products/firefox_desktop/fluent_migrations.md
Comment (Pike) [reply src/products/firefox_desktop/fluent_migrations.md]: `scm_l10n` can be deactivated for the lack of use of hgmo access, which is why the `_active` variants of that matter.

**Comment by flodolo** [comment]
> I'd split the explanations from the process. For example, _don't push migrations from quarantine to gecko-strings outside of this process_ should be the first thing on the docs ;-)

I completely missed this part of the review, let me see what I can do about it.

**Comment by flodolo** [comment]
> I completely missed this part of the review, let me see what I can do about it.

Tried to split the document in two sections, but the reality is that the process takes up 95% of the content, and I think that's OK for the goal of this document (leave instructions for anyone else taking on this work instead of me).

**Comment by flodolo** [comment]
Actually, let me clarify more one bit about Pontoon.

File: src/products/firefox_desktop/fluent_migrations.md
Comment (Pike) [inline src/products/firefox_desktop/fluent_migrations.md]: I kinda know what this is, but can you reword this? It sounds like localizers would be doing sign-offs still, so there'd be pending ones to review. I'd even update the section title.

What do you do when you find issues in the diffs here? Do you just keep track of the locales, or do you wait for the fix to be in the repo?

File: src/products/firefox_desktop/fluent_migrations.md
Comment (Pike) [inline src/products/firefox_desktop/fluent_migrations.md]: I don't think this matters. It used to, but I think the conflicting sync processes is dealt with.

What's missing here is that switching off sync significantly reduces the chance of race conditions between pontoon pushing and a migration.

File: src/products/firefox_desktop/fluent_migrations.md
Comment (Pike) [inline src/products/firefox_desktop/fluent_migrations.md]: As mentioned in my top-level comment, we should know that strings in quarantine are good before starting this.

File: src/products/firefox_desktop/fluent_migrations.md
Comment (Pike) [inline src/products/firefox_desktop/fluent_migrations.md]: I wonder if this should be moved up before you start the whole progress. Maybe into the "quarantine ready" section? I can see that you do this now, as you've built out confidence in the migration initially, but for someone stepping in, this might be unfortunately late in the game. At this point, it's really hard to recover gracefully, right?

File: src/products/firefox_desktop/fluent_migrations.md
Comment (Pike) [inline src/products/firefox_desktop/fluent_migrations.md]: This probably also belongs into the testing section. The sequence of migrations to run is ruled by what you're intending to push from quarantine to gs, there's not really a choice here.

**Review by Pike — CHANGES_REQUESTED** [review]
I've taken another pass at this, with additional attention on the possible target audience. Notably, me and Stas. We'll screw things up differently at different points, so shuffling things that could fail but probably won't is a good idea.

One thing that's in your head that should be explicit for the newbies is the link between quarantine review and migration running.

Can you add a section to `src/products/review/review.md` to watch out for patches with migrations? Or maybe have a follow-up for that?

There should also be a back-reference from `fluent_migrations` to the review doc to make sure that the quarantine is actually good to push. Maybe as a separate section between *Prerequisites* and *Clean up sign-offs*?

I've left some comments inline as well on other aspects that tie migrations and quarantine together.

File: src/products/firefox_desktop/fluent_migrations.md
Comment (flodolo) [reply src/products/firefox_desktop/fluent_migrations.md]: I'm not sure how to rewrite this (I can explicitly call out who's doing the sign off): why do you think it sounds like localizers would be doing sign-offs?

In case of issues, I fix them in Pontoon and wait for the next sync (which is not disabled yet at this point).

File: src/products/firefox_desktop/fluent_migrations.md
Comment (flodolo) [reply src/products/firefox_desktop/fluent_migrations.md]: I'm not sure it's not relevant anymore. If I run the migration for Firefox, but keep Thunderbird enabled, that will still clog Pontoon, because it reads all new changesets in l10n repos. Or do you think that's not the case?

File: src/products/firefox_desktop/fluent_migrations.md
Comment (flodolo) [reply src/products/firefox_desktop/fluent_migrations.md]: Good point. At this point you need to fix the migration or you're screwed. I'll put it before the push to gecko-strings.

File: src/products/firefox_desktop/fluent_migrations.md
Comment (Pike) [inline src/products/firefox_desktop/fluent_migrations.md:76]: I'd leave the `wet-run push` to the all-locales run below, so that it stays in the flow of not touching upstream unless quarantine is OK.

File: src/products/firefox_desktop/fluent_migrations.md
Comment (Pike) [reply src/products/firefox_desktop/fluent_migrations.md]: Rereading this, I think it is less critical here, and more critical when switching sync back on, actually.

Move this down there,  to make sure newbies like me don't just switch sync back on and wait for the scheduled jobs to take it?

**Review by Pike — CHANGES_REQUESTED** [review]
Two nits left, and then there's Sign-offs.

I think it's important that the nomenclature here matches the nomenclature on elmo and on https://mozilla-l10n.github.io/documentation/products/review/signoffs.html. As of now, there's quite a mismatch.

To pinpoint this to something, a pending sign-off is a sign-off that hasn't been accepted, rejected, or obsoleted. Practically, those hardly exist anymore, but they're still part of the documentation implicitly when rejecting a changeset. When you untick the "and accept" checkbox, you create a pending sign-off, and then you review that sign-off and reject it.

There's no existing wording for "stuff you could sign off on" on elmo, or, AFAICT, the sign-offs docs in this repo. They're not-existing sign-offs,  and they can't be pending sign-offs.

Given the state of this patch, I wonder if it'd be better to drop that part from the docs for now, and to try to get to a consistent nomenclature in a follow-up?

File: src/products/firefox_desktop/fluent_migrations.md
Comment (flodolo) [reply src/products/firefox_desktop/fluent_migrations.md]: Good point, moving it.

File: src/products/firefox_desktop/fluent_migrations.md
Comment (flodolo) [reply src/products/firefox_desktop/fluent_migrations.md:76]: I think it should be kept:
- Quarantine is already checked at this point, and Pontoon's sync is disabled.
- You've already committed the changes to the repository in the previous step, because you need them to run `compare-locales`. Here you're just pushing them out without going into the repo and running `hg push`.
- By removing this step, you risk conflicts unnecessarily. By the time the script has reached Italian again, you might have conflicts.

**Review by flodolo — COMMENTED** [review]
> Given the state of this patch, I wonder if it'd be better to drop that part from the docs for now, and to try to get to a consistent nomenclature in a follow-up?

I would really prefer to keep the document as is, so that it's complete, instead of dropping the information. Happy to discuss and fix terminology in a follow-up issue.

To me, *pending sign-offs* are changesets eligible for sign-offs, i.e. there's activity in the repository and it shows up in the Bulk Sign-offs. I used the same terminology (*pending*) there.

**Comment by flodolo** [comment]
> To me, _pending sign-offs_ are changesets eligible for sign-offs, i.e. there's activity in the repository and it shows up in the Bulk Sign-offs. I used the same terminology (_pending_) there.

I'm about to run a new migration:

I check [pending sign-offs here](https://l10n.mozilla.org/shipping/sign-off/fx77) and review everything that's listed in that page (I'm doing sign-offs on Nightly, since Beta is closed). Once the migration is finished, that page will have about 100 sign-offs to take, and I will only check a handful of them, because I trust the migration code, compare-locales, and my own scripts to catch errors. 

Otherwise, the whole migration system would not be sustainable.

**Comment by flodolo** [comment]
> (I'm doing sign-offs on Nightly, since Beta is closed)

BTW, that's something that should be noted in the sign-offs doc in a follow-up, together with a terminology decision around *pending*.

**Comment by Pike** [comment]
To clarify, pending signoffs is a defined concept in elmo:

https://github.com/mozilla/elmo/blob/dad5b93b791c59be109490a363619c7f2892f2f4/apps/shipping/models.py#L125-L131

Which is why, when you're talking to folks that know elmo, using that phrase for something else is very confusing.

**Comment by flodolo** [comment]
> Which is why, when you're talking to folks that know elmo, using that phrase for something else is very confusing.

And I agree that we should define a shared, non ambiguous terminology, but I'd like to decouple that conversation and unblock this work :)

File: src/products/firefox_desktop/fluent_migrations.md
Comment (flodolo) [inline src/products/firefox_desktop/fluent_migrations.md]: The only alternative I can think of is "review pending sign-offs" -> "sign off available changesets". In this part:

> Before starting, sign off in Elmo all available changesets for the current Firefox Beta in the *Bulk Sign-offs* view.

And link to https://mozilla-l10n.github.io/documentation/products/review/signoffs.html#bulk-sign-offs in the last link.

In the next paragraph, remove *pending*.

@Pike 
Would that work?

Comment (Pike) [reply src/products/firefox_desktop/fluent_migrations.md]: Yeah, that would work.

Comment (flodolo) [reply src/products/firefox_desktop/fluent_migrations.md]: Updating this and the sign-off doc to remove mention of *pending*.

I'm keeping "Firefox Beta" here for sake of clarity, but I plan to update the sign-off document in a follow-up, to explain that you can sign off central while Beta is string frozen.

**Review by Pike — APPROVED** [review]
Thanks for hanging in there, r=me.

**Comment by flodolo** [comment]
Thanks! I'll wait next week to send an update to the sign-off docs in your direction :)

**Pull request merged**
